### PR TITLE
Tan11389/nmea location data source

### PR DESCRIPTION
**Pull request by tanneryould**
Adds a sample to demonstrate how to use the NMEA Location Data Source API

### CODE REVIEW

File: ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    // Declare the C++ instance which creates the scene etc. and supply the view
+    DisplayDeviceLocationWithNmeaDataSourcesSample {
+        id: model
+        mapView: view
+    }
+
+    Button {
+        id: button
+        anchors.horizontalCenter: parent.horizontalCenter
+        y: 5
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:38]: couple comments:
* not advisable to use hardcoded x/y values. In this case, you could use `anchors.margins: 5` to achieve the same effect
* I'd opt for the button to be on the bottom. Having at the top is a little distracting and hard to reach with thumb when on a phone

File: ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+The NMEA location data source allows for detailed interrogation of the information coming from the GPS receiver. For example, allowing you to report the number of satellites in view.
+
+## How to use the sample
+Tap "Start" to parse the NMEA sentences into a simulated location data source, and initiate the location display. Tap "Reset" to reset the location display.
+
+## How it works
+
+1. Load NMEA sentences from a local file.
+2. Parse the NMEA sentence strings, and push data into `NMEALocationDataSource`.
+3. Set the `NMEALocationDataSource` to the location display's data source.
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/README.md:20]:
```suggestion
2. Parse the NMEA sentence strings, and push data into `NmeaLocationDataSource`.
3. Set the `NmeaLocationDataSource` to the location display's data source.
```

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                }
+                targetScale: 100000
+            }
+        }
+
+        locationDisplay.dataSource: NmeaLocationDataSource {
+            id: nmeaLocationDataSource
+            receiverSpatialReference: SpatialReference { wkid: 4326 }
+        }
+        locationDisplay.autoPanMode: Enums.LocationDisplayAutoPanModeRecenter;
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:51]: semicolons only at end of javascript code blocks, not properties

```suggestion
        locationDisplay.autoPanMode: Enums.LocationDisplayAutoPanModeRecenter
```

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        }
+    }
+
+    Timer {
+        id: timer
+        interval: 1000
+        repeat: true
+        property var iterator: 0
+        onTriggered: {
+
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:95]:
```suggestion
```

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        property bool sampleStarted: false
+        text: sampleStarted ? "STOP" : "START"
+        onClicked: {
+            sampleStarted = !sampleStarted;
+            if (sampleStarted) {
+                timer.start();
+                nmeaLocationDataSource.start();
+                mapView.locationDisplay.start();
+            }
+            else {
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:68]:
```suggestion
            } else {
```

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                                                              mockNmeaData.push(line + "\n");
+                                                          });
+        }
+    }
+
+    Timer {
+        id: timer
+        interval: 1000
+        repeat: true
+        property var iterator: 0
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:93]: I'd probably call this count or something else with less other meanings.

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            else {
+                timer.stop();
+                timer.iterator = 0;
+                nmeaLocationDataSource.stop();
+                mapView.locationDisplay.stop();
+            }
+        }
+    }
+
+    property var mockNmeaData: []
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:77]: usually we put global properties at the top of the file for visibility/organization (near line 27)

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+    FileFolder {
+        id: mockNmeaDataFile
+        path: ":/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:81]:
```suggestion
        path: "qrc:/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"
```

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            id: nmeaLocationDataSource
+            receiverSpatialReference: SpatialReference { wkid: 4326 }
+        }
+        locationDisplay.autoPanMode: Enums.LocationDisplayAutoPanModeRecenter;
+    }
+
+    Button {
+        id: button
+        anchors.horizontalCenter: parent.horizontalCenter
+        y: 5
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:57]: same comments as with c++

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        width: 200
+        property bool sampleStarted: false
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:59]: same comment as with C++ - different name suggested

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                nmeaLocationDataSource.pushData(mockNmeaData[iterator]);
+                iterator++;
+                if (iterator >= mockNmeaData.length)
+                    iterator = 0;
+                if (mockNmeaData[iterator].startsWith("$GPGGA"))
+                    break;
+            }
+        }
+    }
+}
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:112]: code looks good in the QML API sample, but could use more comments throughout explaining what is happening

File: ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/README.md (hunk 2 above)
Comment (ldanzinger) [reply ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/README.md:20]: There are more throughout this readme and the QML version but I won't mark them all

File: ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/README.metadata.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        "navigation",
+        "NMEA",
+        "NMEALocationDataSource",
+        "real-time",
+        "trace"
+    ],
+    "redirect_from": "",
+    "relevant_apis": [
+        "LocationDisplay",
+        "NMEALocationDataSource"
```
[the remark this answers is not in the document]
Comment (ldanzinger) [reply ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/README.metadata.json:22]: Same issue in the QML version

**Review by ldanzinger — CHANGES_REQUESTED** [review]

File: ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+{
+  // Enable receiving NMEA location data from external device
+  m_nmeaLocationDataSource->start();
+  // Display the user's location
+  m_mapView->locationDisplay()->start();
+
+  // Load simulated NMEA sentences for sample
+  if (m_mockNmeaSentences.isEmpty())
+  {
+    QString filePath = ":/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea";
```
[the remark this answers is not in the document]
Comment (tanneryould) [reply ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp:80]: Weirdly the `qrc:/` prefix breaks it. This runs on an emulated Android as it is

[the remark this answers is not in the document]
Comment (ldanzinger) [reply ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp:80]: What does your path look like with the prefix?

[the remark this answers is not in the document]
Comment (tanneryould) [reply ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp:80]: `"qrc:/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"`

In past samples, `qrc:/` is used as a prefix when creating a `QUrl` but `QFile::setFileName` expects a `const QString &name`.

[the remark this answers is not in the document]
Comment (tanneryould) [reply ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp:80]: `QFile mockDataFile(QUrl("qrc:/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea").toLocalFile());` doesn't work either unfortuantely

[the remark this answers is not in the document]
Comment (ldanzinger) [reply ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp:80]: Ok, what you have is good then

**Comment by tanneryould** [comment]
Thank you for all of your comments. I believe I've addressed everything and cleaned up a few other areas of sloppy coding. 24 requested changes is certainly a low bar to improve on.

File: ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp [hunk 14]
```diff
@@ -101,37 +108,40 @@ void DisplayDeviceLocationWithNmeaDataSources::start()
...
   // Stop displaying user's location
   m_mapView->locationDisplay()->stop();
+
   // Stop receiving location data
   m_nmeaLocationDataSource->stop();
-  m_mapView->locationDisplay()->setAutoPanMode(LocationDisplayAutoPanMode::Off);
 }
 
-bool DisplayDeviceLocationWithNmeaDataSources::loadMockDataFile(QString filePath)
+bool DisplayDeviceLocationWithNmeaDataSources::loadMockDataFile(const QString &filePath)
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp:126]: this is the style we try to use:

```suggestion
bool DisplayDeviceLocationWithNmeaDataSources::loadMockDataFile(const QString& filePath)
```

File: ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.h [hunk 15]
```diff
@@ -50,20 +50,20 @@ class DisplayDeviceLocationWithNmeaDataSources : public QObject
...
 signals:
   void mapViewChanged();
-  void sampleStartedChanged();
+  void nmeaSimulationActiveChanged();
 
 private:
   Esri::ArcGISRuntime::MapQuickView* mapView() const;
   void setMapView(Esri::ArcGISRuntime::MapQuickView* mapView);
-  bool loadMockDataFile(QString filePath);
+  bool loadMockDataFile(const QString &filePath);
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.h:58]:
```suggestion
  bool loadMockDataFile(const QString& filePath);
```

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 16]
```diff
@@ -44,69 +47,94 @@ Rectangle {
...
-        locationDisplay.autoPanMode: Enums.LocationDisplayAutoPanModeRecenter;
+        locationDisplay.autoPanMode: Enums.LocationDisplayAutoPanModeRecenter
     }
 
     Button {
         id: button
         anchors.horizontalCenter: parent.horizontalCenter
-        y: 5
+        anchors.bottom: parent.bottom
+        anchors.margins: 25
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:62]: anchors can be grouped:

```suggestion
        anchors {
            bottom: parent.bottom
            margins: 25
        }
```

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml [hunk 17]
```diff
@@ -44,69 +47,94 @@ Rectangle {
...
+                endSimulation();
         }
     }
 
-    property var mockNmeaData: []
-
     FileFolder {
         id: mockNmeaDataFile
-        path: ":/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"
+        path: "://Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"
```
Comment (ldanzinger) [inline ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:76]: is this change intentional?

**Review by ldanzinger — APPROVED** [review]
Looks good - have a couple more formatting changes suggested, but no need to re-review before merging.

File: ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml (hunk 17 above)
Comment (tanneryould) [reply ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:76]: Good catch. It's not. It's left over from testing how I can include `qrc` as we discussed. It still runs, but it should be removed.
